includes/admin/class-llms-sendwp.php

```diff
@@ -0,0 +1,399 @@
+		ob_clean();
+		wp_send_json( $ret, ! empty( $ret['status'] ) ? $ret['status'] : 200 );
+
+	}
+
+	/**
+	 * Remote installation method.
+	 *
+	 * @since [version]
+	 *
+	 * @return array
+	 */
+	public function do_remote_install() {
+
+		if ( ! current_user_can( 'install_plugins' ) ) {
+			return array(
+				'code'    => 'llms_sendwp_install_unauthorized',
+				'message' => __( 'You do not have permission to perform this action.', 'lifterlms' ),
+				'status'  => 403,
+			);
+		}
+
+		$install = $this->install();
+		if ( is_wp_error( $install ) ) {
+			return array(
+				'code'    => $install->get_error_code(),
+				'message' => $install->get_error_message(),
+				'status'  => 400,
+			);
+		}
+
+		return array(
+			'partner_id'      => 2007,
+			'register_url'    => sendwp_get_server_url() . '_/signup',
+			'client_name'     => sendwp_get_client_name(),
+			'client_secret'   => sendwp_get_client_secret(),
+			'client_redirect' => sendwp_get_client_redirect(),
+		);
+
+	}
+
+	/**
+	 * Install / Activate SendWP plugin.
+	 *
+	 * @since [version]
+	 *
+	 * @return WP_Error|true
+	 */
+	private function install() {
+
+		$is_sendwp_installed = false;
+		foreach ( get_plugins() as $path => $details ) {
+			if ( false === strpos( $path, '/sendwp.php' ) ) {
+				continue;
+			}
+			$is_sendwp_installed = true;
+			$activate            = activate_plugin( $path );
+			if ( is_wp_error( $activate ) ) {
+				return $activate;
+			}
+			break;
+		}
+
+		if ( ! $is_sendwp_installed ) {
+
+			include_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+			include_once ABSPATH . 'wp-admin/includes/file.php';
+			include_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+
+			// Use the WordPress Plugins API to get the plugin download link.
+			$api = plugins_api(
+				'plugin_information',
+				array(
+					'slug' => 'sendwp',
+				)
+			);
+			if ( is_wp_error( $api ) ) {
+				return $api;
+			}
+
+			// Use the AJAX upgrader skin to quietly install the plugin.
+			$upgrader = new Plugin_Upgrader( new WP_Ajax_Upgrader_Skin() );
+			$install  = $upgrader->install( $api->download_link );
+			if ( is_wp_error( $install ) ) {
+				return $install;
+			}
+
+			$activate = activate_plugin( $upgrader->plugin_info() );
+			if ( is_wp_error( $activate ) ) {
+				return $activate;
+			}
+		}
+
+		// Final check to see if SendWP is available.
+		if ( ! function_exists( 'sendwp_get_server_url' ) ) {
+			return new WP_Error( 'llms_sendwp_not_found', __( 'SendWP Plugin not found. Please try again.', 'lifterlms' ), $install );
+		}
+
+		return true;
+
+	}
+
+	/**
+	 * Find the end of the "email_options" section to splice in new settings.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $settings Default engagement settings.
+	 * @return int
+	 */
+	private function get_splice_index( $settings ) {
+		foreach ( $settings as $i => $setting ) {
+			if ( 'email_options' === $setting['id'] && 'sectionend' === $setting['type'] ) {
+				return $i + 1;
+				break;
```

Comment:
if we `return` we don't need a `break`, right?

Reply:
no, no we don't.